doc/rest_api/snippet_sharing/api-description.apib

```diff
@@ -0,0 +1,410 @@
+
+
+# API Description [/]
+
+## List API methods [GET]
+
+Returns an overview of available API methods with required parameters and a description.
+
++ Response 200 (text/html; charset=utf-8)
+
+
+
+# Authentication endpoint [/auth]
+
+## Authenticate [POST]
+
+The API will try to authenticate the user by the submitted credentials. In case of success a session token (JWT) will be returned, which can be used for further requests to the API.
+
++ Request (application/x-www-form-urlencoded)
+    + Attributes
+        + username: the_username (string, required)
+        + password: aSecretPassword (string, required)
+        
+    + Schema
+    
+            {
+                "type": "object",
+                "properties": {
+                    "username": {
+                        "description": "Name of the user that wants to authenticate",
+                        "type": "string"
+                    },
+                    "password": {
+                        "description": "Password for the username in question",
+                        "type": "string"
+                    }
+                },
+                "required": [
+                    "username",
+                    "password"
+                ],
+                "additionalProperties": false
+            }
+ 
++ Response 200 (application/json)
+    + Body
+
+            {
+                "status": 200,
+                "token": "eyJ...Q2tY="
+            }
+            
+    + Schema
+    
+            {
+                "type": "object",
+                "properties": {
+                    "status": {
+                        "description": "HTTP status code",
+                        "type": "integer",
+                        "minimum": 0
+                    },
+                    "token": {
+                        "description": "Session token to be used as authentication for further requests.",
+                        "type": "string"
+                    }
+                },
+                "required": [
+                    "status",
+                    "token"
+                ],
+                "additionalProperties": false
+            }
+
++ Response 401 (application/json)
+    + Body
+    
+            {
+                "status": 401,
+                "error": "Unauthorized",
+                "message": "You have to supply an username and a password.",
+                "loca": "AUTH_NEED_USERNAME_PASSWORD"
+            }
+            
+    + Schema
+
+            {
+                "type": "object",
+                "properties": {
+                    "status": {
+                        "description": "HTTP status code",
+                        "type": "integer",
+                        "minimum": 0
+                    },
+                    "error": {
+                        "description": "Textual description of HTTP status code",
```

Comment:
> Why not use english text as error and translate that text (without error codes)?

Because this creates limitations for other software. For example is the frontend loca plugin I use not capable of translating strings with spaces in it. It is also more likely that the spelling/wording of a detailed error message changes than a simple i18n key.

> It should not only be a hint, but when e.g. the password is too short there should be an error telling exactly this, not a list of possible problems.

There is a huge trade-off coming with this requirement. I thought a lot about it when writing the code to be honest. 
Basically there are two options: 
- Check every single constraint on its own (which would lead to many many checks for some inputs).
- Or put some checks together and see if it makes sense. 

For now I've chosen the second approach, because I don't really see where it could lead to problems. If the message says that three fields are required and the user only submitted two of them, it should not be a big deal to find out what the problem is. Other error messages state very exactly what the problem is, for example this one

> The given username is not valid. The username must be between 3 and 20 signs long and may contain only letters a-z (lower- and upper-case), the numbers 0-9 and dots (.) as well as dashes (-).